### ISSUE TITLE
Make task queues more ordered

### ISSUE THREAD

**Issue by mfateev** [temporalio/temporal#2517]
**Is your feature request related to a problem? Please describe.**
Currently, task queue sync match is prioritized over backlog. So a task can sit in a backlog for a long time being starved by new tasks. Ideally task queues should try to deliver tasks in order. It is not possible to guarantee total ordering as multiple partitions and worker processes are involved, but not starving backlogged requests would greatly improve fairness of the system.

**Describe the solution you'd like**
Deliver tasks from backlog first. It can increase DB utilization as all tasks would need to be persisted in the case of a backlog. So we should consider if this should be an option of a task queue. I would try to avoid this and introduce such an option only if we see that DB load increase is significant after performance testing.

**Comment by dnr** [comment]
Sync match isn't prioritized over backlog right now. They both write into one go channel. I suppose the ratio will be tilted toward sync match since there could be multiple sync match writers and only one db writer, but backlog tasks should still get through. We talked about various potential changes to this last year and decided to wait until we had time to address all the concerns more holistically (avoiding starvation, ordering, avoiding increasing load further when load is already elevated).

**Comment by mfateev** [comment]
I think we should have an option to not perform sync match if there is a backlog.

**Comment by ShahabT** [comment]
Starting to work on this. As the first step, I'm thinking to add a boolean dynamic config (per namespace+task queue) to allow disabling sync match when there is a backlog. Intention is to manually set this config only when the specific task queue requires FIFOish ordering.

**Comment by paulnpdev** [comment]
what do you mean by "manually"?  Is the intent that a customer can
programmatically set this?

**Comment by ShahabT** [comment]
@paulnpdev no, this config will not be exposed to the client (at least not now). Only the server operator can turn this flag on. Because this has implications on the load on the persistence layer, I do not think we can allow the customers turning it on, until the dynamics are fully understood as you mentioned.

**Comment by paulnpdev** [comment]
ok for now!
will the flag be at the namespace level, the WF type level, or the TaskQ
level

**Comment by ShahabT** [comment]
@paulnpdev taskQ level. More specifically, the combination of namespace + taskQ name + taskQ type (workflow vs activity)

**Comment by ShahabT** [comment]
After deeper analysis based on synthetic and customer workloads, it turned out that:
- **Sync match rarely happens in presence of a backlog.** This is because sync match request (for tasks coming directly from history) only go through if a poller is available in that very moment, they do not wait and compete with the pending backlog task for new pollers to become available.
- **Task queue partitioning is the main reason for dispatch order mixups.** Due to the asymmetricity of our root vs leaf partitions, when there is a backlog, the root partition tends to process it's backlog much faster and the leaf partitions accumulate heavier/older backlogs. [I'm investigating how/why this happens]

Good News
-----------
For task queues with low/medium load (in terms of tasks per sec) where a single task queue partition can do all the work **you can get approximate FIFO ordering by reducing the number of partitions to 1**. The following dynamic configs should be used for that:
```
matching.numTaskqueueWritePartitions:
  - value: 1
    constraints:
      namespace: <namespace name>
      taskQueueName: <task queue name>
matching.numTaskqueueReadPartitions:
  - value: 1
    constraints:
      namespace: <namespace name>
      taskQueueName: <task queue name>
```
Note that if you are changing this prop for a pre-exsiting task queue, `matching.numTaskqueueWritePartitions` should be reduced first, and after making sure no unfinished tasks are in the old partitions, you can update `matching.numTaskqueueReadPartitions` too.

**Comment by macnibblet** [comment]
@ShahabT How do we do this in the temporal cloud?

**Comment by kshitij-g** [comment]
Is there any update on this issue?

**Comment by ShahabT** [comment]
@awartoft in Cloud you can create a ticket and ask support to set the number of partitions to your (low traffic) task queue to 1 so you can benefit from the fifo behavior.

@paulnpdev I agree that if the sync-match task belongs to high a high priority or a fairness key that should not be backlogged (or has tiny backlog), we should avoid DB write/read. I believe @dnr has optimization ideas here.

@kshitij-g there is no further update here, I consider this issue completed and closing it now. 

TLDR: as of now, we provide exact FIFO for small task queues (single-partition) and for larger ones the behavior is FIFOish because by design, the backlog in partitions remain pretty balanced.

(in the future, partitioning will be more dynamic and changing number of partitions may bring unbalances, but we might add more coordination between partitions in that case to preserve FIFOish-ness, TBD)
